### DISCUSSION TITLE
OTLP Exporter to Grafana Alloy - Impossible to Trace/Debug on Mac with Rider

### DISCUSSION THREAD

**Discussion by aussiearef** [open-telemetry/opentelemetry-dotnet#5536] [Q&A]
I have the minimal web API code below with OpenTelemety and OTLP EXporter. I am trying to send a simple metric over to an OTEL Collector called Grafana Alloy and then forward it to Prometheus.

I don't get the metric in Prometheus, and Alloy's internal metrics indicate that everything still needs to be received by its receivers. So I am now wondering if OTLP Exporter is sending anything to Alloy.

The code is below:

```
Using System.Diagnostics.Metrics;
using OpenTelemetry;
using OpenTelemetry.Exporter;
using OpenTelemetry.Metrics;
using OpenTelemetry.Resources;

var builder = WebApplication.CreateBuilder(args);
var collectorUrl = builder.Configuration["OtelCollector:Host"]?? "http://localhost:4318/v1/metrics";

Builder.Services.AddOpenTelemetry()
    .WithMetrics(m =>
{
    m.SetResourceBuilder(ResourceBuilder.CreateDefault()
            .AddService("OrderService"))
        .AddMeter("OrderServiceMeter")
        .AddInstrumentation<HttpClient>()
        .AddOtlpExporter(o =>
        {
            o.Protocol = OtlpExportProtocol.HttpProtobuf;
            o.Endpoint = new Uri(collectorUrl);
        })
        .AddConsoleExporter();
    
});
Builder.Services.AddMetrics();

var app = builder.Build();


Counter<int> otelOrderCount= null;

app.Use(async (context, next) =>
{
    var metricFactory = context.RequestServices.GetService<IMeterFactory>();
    var meter = metricFactory?.Create(new MeterOptions("OrderServiceMeter"));
    otelOrderCount = meter.CreateCounter<int>("otel_order_count");
    otelOrderCount?.Add(1);
    await next();
});

app.MapGet("/", () =>
{
    otelOrderCount?.Add(1);
    return "OK";
});

app.Run();

```

I have included the OTEL_DIAGNOSTICS.json file in my project (copied over to /bin/debug/net8.0 directory) when I run "dotnet trace" a .nettrace file is generated. Then I convert it to speedscope format, and it gives me a JSON file with no trace data in it.

It seems that assuming that every .NET developer does the development and testing on a Windows machine and being overly reliant on dotnet trace has very much limited the debugging and traceability of the OTLP Exporter library!

My question is:

How can I make sure that OTLP Exporter is actually exporting the metrics to OTEL Collector?! 
How can I see HTTP failure or even the HTTP Responses coming back from the OTEL Collector (in my case Grafana Alloy)?!

Is it possible to emit the debug and trace info on the stdout or Console without going through the pain of using dotnet trace on a a Mac computer?

p.s. I use Jetbrain Rider.
p.s. ConsoleExporter correctly writes the metric (otel_order_count) in the Console.

**Comment by pyohannes** [comment]
Please follow the steps described [here](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/README.md#self-diagnostics), make sure that the file `OTEL_DIAGNOSTICS.json` is placed in the current working directory of your application.

After running your application, you then get a file named `<executable name>.<process id>.log` in the directory specified in `OTEL_DIAGNOSTICS.json`, this file should contain helpful error messages. 

Please post those error messages here if you think the problem lies with the .NET SDK, otherwise I'd recommend [reaching out to Alloy folks](https://github.com/grafana/alloy?tab=readme-ov-file#community).
